Prepared statement - Frequent crash on ProxySQL

Hi @renecannao ,

ProxySQL is crashing every now and then, I've processed coredumps and each core dump is pointing to below line in MySQL_Session.cpp.

I've tested with 1.4.16,2.0.14 and even 2.1.1. All versions having the same crash, 

![imageedit_6_4852359767](https://user-images.githubusercontent.com/15307959/112464484-908b0d00-8d89-11eb-9cf8-a07b670bcb85.png)


Errorlog:
```
proxysql: MySQL_Session.cpp:3297: int MySQL_Session::handler(): Assertion `0' failed.
Error: signal 6:
/usr/bin/proxysql(_Z13crash_handleri+0x1a)[0x4d5fda]
/lib64/libc.so.6(+0x35270)[0x7f9262f09270]
/lib64/libc.so.6(gsignal+0x37)[0x7f9262f091f7]
/lib64/libc.so.6(abort+0x148)[0x7f9262f0a8e8]
/lib64/libc.so.6(+0x2e266)[0x7f9262f02266]
/lib64/libc.so.6(+0x2e312)[0x7f9262f02312]
/usr/bin/proxysql(_ZN13MySQL_Session7handlerEv+0x395f)[0x5386cf]
/usr/bin/proxysql(_ZN12MySQL_Thread20process_all_sessionsEv+0x3cc)[0x51695c]
/usr/bin/proxysql(_ZN12MySQL_Thread3runEv+0xc4c)[0x51d68c]
/usr/bin/proxysql(_Z24mysql_worker_thread_funcPv+0x6c)[0x4ceeec]
/lib64/libpthread.so.0(+0x7e25)[0x7f92640e0e25]
/lib64/libc.so.6(clone+0x6d)[0x7f9262fcc34d]
2021-03-25 09:22:41 main.cpp:1573:ProxySQL_daemonize_phase3(): [ERROR] ProxySQL crashed. Restarting!
2021-03-25 09:22:41 [INFO] ProxySQL version 2.0.14-70-g7b02b19
```

```
#0  0x00007f9262f091f7 in raise () from /lib64/libc.so.6
No symbol table info available.
#1  0x00007f9262f0a8e8 in abort () from /lib64/libc.so.6
No symbol table info available.
#2  0x00007f9262f02266 in __assert_fail_base () from /lib64/libc.so.6
No symbol table info available.
#3  0x00007f9262f02312 in __assert_fail () from /lib64/libc.so.6
No symbol table info available.
#4  0x00000000005386cf in MySQL_Session::handler (this=this@entry=0x7f925192a300) at MySQL_Session.cpp:3297
        client_stmt_id = <optimized out>
        stmt_global_id = <optimized out>
        stmt_info = <optimized out>
        stmt_meta_found = <optimized out>
        stmt_meta = 0x7f9200000000
        prepared_stmt_with_no_params = false
        pkt = {ptr = 0x7f925e652940, size = 15}
        c = <optimized out>
        __PRETTY_FUNCTION__ = "int MySQL_Session::handler()"
        __func__ = "handler"
        wrong_pass = false
#5  0x000000000051695c in MySQL_Thread::process_all_sessions (this=this@entry=0x7f925e5fe000) at MySQL_Thread.cpp:4396
        sess = 0x7f925192a300
        total_active_transactions_ = 0
        idle_maintenance_thread = false
        sess_sort = <optimized out>
        __func__ = "process_all_sessions"
        n = 0
        rc = <optimized out>
#6  0x000000000051d68c in MySQL_Thread::run (this=this@entry=0x7f925e5fe000) at MySQL_Thread.cpp:4118
        num_idles = <optimized out>
        ttw = <optimized out>
        maintenance_interval = <optimized out>
        idle_maintenance_thread = <optimized out>
        __PRETTY_FUNCTION__ = "void MySQL_Thread::run()"
        __func__ = "run"
        n = <optimized out>
        rc = <optimized out>
#7  0x00000000004ceeec in mysql_worker_thread_func (arg=0x7f9261c8a440) at main.cpp:760
        thread_attr = {__size = '\000' <repeats 17 times>, "\020", '\000' <repeats 37 times>, __align = 0}
        tmp_stack_size = 8388608
---Type <return> to continue, or q <return> to quit---
        mysql_thread = 0x7f9261c8a440
        worker = 0x7f925e5fe000
#8  0x00007f92640e0e25 in start_thread () from /lib64/libpthread.so.0
No symbol table info available.
#9  0x00007f9262fcc34d in clone () from /lib64/libc.so.6
```

If needed, I can send the full core dumps. Any help is appreciated as we are facing severe stability issues.

I encounter the same problem...


proxysql: MySQL_Session.cpp:3690: int MySQL_Session::handler(): Assertion `0' failed.
Error: signal 6:
/usr/bin/proxysql(_Z13crash_handleri+0x1a)[0x4ae55a]
/lib64/libc.so.6(+0x363b0)[0x7fbd94f603b0]
/lib64/libc.so.6(gsignal+0x37)[0x7fbd94f60337]
/lib64/libc.so.6(abort+0x148)[0x7fbd94f61a28]
/lib64/libc.so.6(+0x2f156)[0x7fbd94f59156]
/lib64/libc.so.6(+0x2f202)[0x7fbd94f59202]
/usr/bin/proxysql(_ZN13MySQL_Session7handlerEv+0x3957)[0x512bf7]
/usr/bin/proxysql(_ZN12MySQL_Thread20process_all_sessionsEv+0x2fc)[0x4f009c]
/usr/bin/proxysql(_ZN12MySQL_Thread3runEv+0xbb4)[0x4f72c4]
/usr/bin/proxysql(_Z24mysql_worker_thread_funcPv+0x6c)[0x4a791c]
/lib64/libpthread.so.0(+0x7e65)[0x7fbd96140e65]
/lib64/libc.so.6(clone+0x6d)[0x7fbd9502888d]
2022-02-28 10:23:50 main.cpp:1538:ProxySQL_daemonize_phase3(): [ERROR] ProxySQL crashed. Restarting!
2022-02-28 10:23:50 [INFO] ProxySQL version 2.0.10-27-g5b31997
2022-02-28 10:23:50 [INFO] ProxySQL SHA1 checksum: e1982906ea874ad66ba381b267abd085720dd934
2022-02-28 10:23:50 [INFO] Angel process started ProxySQL process 3580

--- a/lib/MySQL_Session.cpp
+++ b/lib/MySQL_Session.cpp
@@ -3790,10 +3790,13 @@ void MySQL_Session::handler___status_WAITING_CLIENT_DATA___STATE_SLEEP___MYSQL_C
 		CurrentQuery.stmt_client_id=client_stmt_id;
 		stmt_global_id=client_myds->myconn->local_stmts->find_global_stmt_id_from_client(client_stmt_id);
 		if (stmt_global_id == 0) {
-			// FIXME: add error handling
-			// LCOV_EXCL_START
-			assert(0);
-			// LCOV_EXCL_STOP
+			l_free(pkt.size,pkt.ptr);
+			client_myds->setDSS_STATE_QUERY_SENT_NET();
+			string err_msg = "Unknown prepared statement handler (" + to_string(client_stmt_id) + ") given to mysql_stmt_precheck";
+			client_myds->myprot.generate_pkt_ERR(true,NULL,NULL,1,1243,(char *)"HY000", err_msg.c_str());
+			client_myds->DSS=STATE_SLEEP;
+			status=WAITING_CLIENT_DATA;
+			return;
 		}
 		CurrentQuery.stmt_global_id=stmt_global_id;
 		// now we get the statement information
@@ -3803,7 +3806,8 @@ void MySQL_Session::handler___status_WAITING_CLIENT_DATA___STATE_SLEEP___MYSQL_C
 			// we couldn't find it
 			l_free(pkt.size,pkt.ptr);
 			client_myds->setDSS_STATE_QUERY_SENT_NET();
-			client_myds->myprot.generate_pkt_ERR(true,NULL,NULL,1,1045,(char *)"28000",(char *)"Prepared statement doesn't exist", true);
+			string err_msg = "Unknown prepared statement handler (" + to_string(client_stmt_id) + ") given to mysql_stmt_precheck";
+			client_myds->myprot.generate_pkt_ERR(true,NULL,NULL,1,1243,(char *)"HY000", err_msg.c_str());
 			client_myds->DSS=STATE_SLEEP;
 			status=WAITING_CLIENT_DATA;
 			return;

--- a/test/tap/tests/reg_test_3371_prepared_statement_crash-t.cpp
+++ b/test/tap/tests/reg_test_3371_prepared_statement_crash-t.cpp
@@ -1,0 +1,89 @@
+/**
+ * @file reg_test_3371_prepared_statement_crash-t.cpp
+ * @brief Tries to execute prepared statements with a not existing stmt_id.
+ * This used to crash ProxySQL , so this tap test verifies that ProxySQL
+ * doesn't crash
+ */
+
+
+#include <string>
+#include <stdio.h>
+#include <cstring>
+#include <unistd.h>
+
+#include "mysql.h"
+
+#include "tap.h"
+#include "command_line.h"
+#include "utils.h"
+
+using std::string;
+
+const int NUM_LOOPS = 100; ///< Number of loops for statement execution.
+
+int main(int argc, char** argv) {
+	CommandLine cl;
+
+	// Checking for required environmental variables
+	if (cl.getEnv()) {
+		diag("Failed to get the required environmental variables.");
+		return -1;
+	}
+
+	plan(1+NUM_LOOPS*2); // Plan for testing purposes
+
+	MYSQL* mysql = mysql_init(NULL); ///< MySQL connection object
+	if (!mysql) {
+		fprintf(stderr, "File %s, line %d, Error: %s\n", __FILE__, __LINE__, mysql_error(mysql));
+		return exit_status();
+	}
+
+	// Connecting to ProxySQL
+	diag("Connecting to '%s@%s:%d'", cl.mysql_username, cl.mysql_host, cl.port);
+	if (!mysql_real_connect(mysql, cl.mysql_host, cl.mysql_username, cl.mysql_password, NULL, cl.port, NULL, 0)) {
+		fprintf(stderr, "File %s, line %d, Error: %s\n", __FILE__, __LINE__, mysql_error(mysql));
+		return exit_status();
+	}
+
+	// Initialize and prepare all the statements
+	MYSQL_STMT* stmt = mysql_stmt_init(mysql);
+		if (!stmt) {
+			fprintf(stderr, "mysql_stmt_init(), out of memory\n");
+			return exit_status();
+		}
+
+	std::string select_query = "SELECT 1";
+	diag("select_query: %s", select_query.c_str());
+	if (mysql_stmt_prepare(stmt, select_query.c_str(), strlen(select_query.c_str()))) {
+		fprintf(stderr, "mysql_stmt_prepare at line %d failed: %s\n", __LINE__ , mysql_error(mysql));
+		mysql_close(mysql);
+		mysql_library_end();
+		return exit_status();
+	}
+
+	diag("Increasing stmt_id by 1, so that mysql_stmt_execute() must fail");
+	int rc = 0;
+	for (int i = 0; i < NUM_LOOPS ; i++) {
+		stmt->stmt_id += 1;
+		rc = mysql_stmt_execute(stmt);
+		ok (rc , "mysql_stmt_execute() must fail");
+		if (rc) {
+			unsigned int psrc = mysql_stmt_errno(stmt);
+			ok( psrc == 1243 , "mysql_stmt_execute at line %d failed: %d , %s", __LINE__ , psrc , mysql_stmt_error(stmt));
+		}
+	}
+
+	diag("Decreasing stmt_id by 1, so that mysql_stmt_execute() must succeed");
+	stmt->stmt_id -= NUM_LOOPS;
+	rc = mysql_stmt_execute(stmt);
+	ok (rc == 0 , "mysql_stmt_execute() succeeded");
+	if (rc) {
+		fprintf(stderr, "mysql_stmt_execute at line %d failed: %d , %s\n", __LINE__ , rc , mysql_stmt_error(stmt));
+		mysql_close(mysql);
+		mysql_library_end();
+		return exit_status();
+	}
+	mysql_close(mysql);
+
+	return exit_status();
+}

